lit_gpt/lora.py

```diff
@@ -133,11 +133,38 @@ def reset_parameters(self) -> None:
             nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
             nn.init.zeros_(self.lora_B)
 
+    def get_lora_AB(self) -> torch.Tensor:
+        """Return merged lora_A and lora_B matrices with the same shape as the pretrained weights."""
+        return (self.lora_B @ self.lora_A) * self.scaling
+
     def merge(self) -> None:
         """Merges the LoRA weights into the full-rank weights (W = W + delta_W)."""
         if self.r > 0 and not self.merged:
-            # Merge the weights and mark it
-            self.linear.weight.data += (self.lora_B @ self.lora_A) * self.scaling
+            pretrained_dtype = self.linear.weight.data.dtype
+            lora_data = self.get_lora_AB()
+            # if the pretrained weights and LoRA weights are of the same dtype - simply sum them
+            if pretrained_dtype == lora_data.dtype:
+                self.linear.weight.data += lora_data
+            # if only the pretrained are in quantized form - dequantize, sum with LoRA and quantize the result
+            elif pretrained_dtype == torch.uint8:
+                import bitsandbytes as bnb
+
+                weight = self.linear.weight
+                # capture args like `compress_statistics`, `quant_type` and `quant_state`
+                weight_kwargs = weight.__dict__
+                # dequantize the pretrained weights and sum them with LoRA weights
+                weight_data = bnb.functional.dequantize_4bit(weight.data, weight.quant_state) + lora_data
+                # weights are quantized when they are moved to CUDA device,
+                # so we have to first move them to CPU and after - to CUDA
+                self.linear.weight = bnb.nn.Params4bit(weight_data.to("cpu"), requires_grad=False, **weight_kwargs).to(
+                    weight.device
+                )
```

Comment:
Why did you choose to

dequanize param -> add param and weight -> quantize param

instead of 

quantize weight -> add param and weight

Isn't the latter faster and simpler?

Reply:
This process is even less efficient than I initially thought.
With this approach the weights will be always dequantized to `float16` and, if `lora_data` is in `bfloat16`, the summation will be in `float32`. Which makes sense, since PyTorch has to use a common dtype for both values and the cast to such a dtype should not lead to loss in a precision. Hence `float32`.

> Or is it because you fear that there would be too much of a precision loss?

At the moment of creation of this PR, I wasn't sure what led to the “strange” shape of the matrix, so decided that the summation of dequantized values is the safest option. And I wanted to have something that works, maybe not in the most efficient way, just to solve the issue, and then to work on optimizations.
Plus, as I understand, the data in `quant_state` that helps to dequantize without losses, might be not relevant after we sum two tensors.

Overall, I like your suggestion more, but I need to do more research. There is a lot of stuff that is done implicitly. I'm not the smartest person, so might miss something. I'm not Rick Sanchez, more like Morty :)

Reply:
Each `quant_state` contains at least one `absmax` value (two if we use double quantization).
Since this `absmax` value is heavily used for rescaling of data through normalization by the absolute maximum, it's important to keep it sorta “up to date”.

So far, I don't know if it's even feasible to update `absmax` value/values of the sum of two quantized tensors without dequantizing them first.

So I would stick to the `dequantize->sum->quantize` approach, only add some optimizations to it:
- Cast dequantized value to the dtype of the other term of summation, so we don't have an intermediate cast to torch.float32 (e.g., dequantized is float16 and lora_data is bfloat16)
- Put quantized values directly into the layer on a GPU without moving it on a CPU first and then back to GPU